test/FastaToAssembly_unit_test.py

```diff
@@ -624,4 +622,35 @@ def test_import_fasta_mass_fail_min_contig_length():
         ('If provided, min_contig_length must be an integer, got: {}',
           _update(b, {'min_contig_length': {}})),
     ]
-    _run_test_spec_fail(test_spec, mass=True)
\ No newline at end of file
+    _run_test_spec_fail(test_spec, mass=True)
+
+
+@patch('AssemblyUtil.FastaToAssembly._get_serialized_object_size')
+def test_assembly_objects_generator_with_big_serialized_object_size(mocked_function):
+    fta, _ = _set_up_mocks()
+    assembly_objects = [
+        {'md5': '1e007bad0811a6d6e09a882d3bf802ab', 'base_counts': {'A': 1200415, 'G': 852652, 'C': 846697, 'T': 1191689, 'N': 847}, 'dna_size': 4092300, 'gc_content': 0.41526},
+        {'md5': 'a26f200923f8c860f86a8d728055fd02', 'base_counts': {'A': 1184671, 'G': 847148, 'C': 850586, 'T': 1196308, 'N': 218}, 'dna_size': 4078931, 'gc_content': 0.41622},
+        {'md5': '8aa6b1244e18c4f93bb3307902bd3a4d', 'base_counts': {'A': 1199516, 'G': 850011, 'C': 845769, 'T': 1183802, 'N': 106}, 'dna_size': 4079204, 'gc_content': 0.41571}
+    ]
+    assembly_names = ["name_1", "name_2", "name_3"]
+    mocked_function.return_value = _MAX_DATA_SIZE * 1/3
+    res = list(fta._assembly_objects_generator(assembly_objects, assembly_names))
+    assert len(res) == 2
+    assert res[0][1] == ["name_1", "name_2"]
+    assert res[1][1] == ["name_3"]
+
+
+@patch('AssemblyUtil.FastaToAssembly._get_serialized_object_size')
+def test_assembly_objects_generator_with_small_serialized_object_size(mocked_function):
+    fta, _ = _set_up_mocks()
+    assembly_objects = [
+        {'md5': '1e007bad0811a6d6e09a882d3bf802ab', 'base_counts': {'A': 1200415, 'G': 852652, 'C': 846697, 'T': 1191689, 'N': 847}, 'dna_size': 4092300, 'gc_content': 0.41526},
+        {'md5': 'a26f200923f8c860f86a8d728055fd02', 'base_counts': {'A': 1184671, 'G': 847148, 'C': 850586, 'T': 1196308, 'N': 218}, 'dna_size': 4078931, 'gc_content': 0.41622},
+        {'md5': '8aa6b1244e18c4f93bb3307902bd3a4d', 'base_counts': {'A': 1199516, 'G': 850011, 'C': 845769, 'T': 1183802, 'N': 106}, 'dna_size': 4079204, 'gc_content': 0.41571}
+    ]
+    assembly_names = ["name_1", "name_2", "name_3"]
+    mocked_function.return_value = 9
+    res = list(fta._assembly_objects_generator(assembly_objects, assembly_names))
+    assert len(res) == 1
+    assert res[0][1] == assembly_names
```

Comment:
https://github.com/kbaseapps/AssemblyUtil/pull/89#discussion_r1326545571

I had 2 unit tests for _assembly_objects_generator. One with overflow, one without.

Reply:
I thought that was superceded by https://github.com/kbaseapps/AssemblyUtil/pull/89#issuecomment-1721733074

The fact that the unit tests are calling private methods points to needing integration tests to test the functionality as it would be exercised via the API

Reply:
Got all mixed up. 

https://github.com/kbaseapps/AssemblyUtil/pull/89#issuecomment-1721733074 is to test parallelization of _import_fasta_mass function. I did it here: https://github.com/kbaseapps/AssemblyUtil/blob/c21156573dae25a308d90ac8673c9fbd1c65ea18/test/save_assemblies_from_fastas_integration_test.py#L582

Reply:
yeah, and I think you could probably alter that slightly (and maybe add a batch_size param to the API to make this easy to test) and test the last chunk from the generator that way and you wouldn't need the unit tests running against private methods

Reply:
Added max_cumsize param

Reply:
I'm saying you should do this in the integration test

Reply:
Let's say we do this in the integration test and we can pass a very small max_cumsize param to the API through import_fasta_mass() function, making it overflow easily. 

But I am curious how would we test assembly_objects_generator()?  The result it yields will be extended into assembly_infos.

Reply:
You're not testing the method directly, you're testing the code in the method by exercising it via the API. This allows us to change the batching (which is an implementation detail) without having to rewrite a ton of tests. 

If you really really want to make sure the batch size is as expected, the way to do that would be to turn off parallelzation, mock DFU, and make sure the batch sizes were as expected. That seems like overkill to me though

Reply:
Compiler did not preserve the new param I passed in though.

save_assemblies_from_fastas(self, ctx, params, max_cumsize=None)

Reply:
the compiler compiles the impl file from the spec file. No changes to spec file = no changes to functions.

That being said, that param shouldn't be part of the modules public API, which is what the spec defines. For this case you should do the integration test at the FastaToAssembly level.

Reply:
Updated!

Reply:
Codecov says those lines are still not covered. Is the test correct and is it getting run?

Reply:
yeah, I tested generator overflow in the integration test at the FastaToAssembly level. 

6 assembly files with 2 workers. 3 files per worker.

['GCF_000970185.1_ASM97018v1_genomic.fna.gz', 'GCF_000970205.1_ASM97020v1_genomic.fna.gz',
'GCF_000970245.1_ASM97024v1_genomic.fna.gz'] is in one batch as they are all tiny.

['GCA_002506415.1_ASM250641v1_genomic.fna.gz', 'GCF_000007065.1_ASM706v1_genomic.fna.gz']
['GCF_000970165.1_ASM97016v1_genomic.fna.gz'] into 2 batch as 'GCF_000007065.1_ASM706v1_genomic.fna.gz is huge. 

<img width="1439" alt="Screen Shot 2023-10-06 at 10 55 23 AM" src="https://github.com/kbaseapps/AssemblyUtil/assets/22202096/3240a5fa-db91-4794-b9da-396f758e4147">

Reply:
If you make the block that codecov says is untested throw an exception, does your test fail?

Reply:
<img width="1439" alt="Screen Shot 2023-10-06 at 1 29 06 PM" src="https://github.com/kbaseapps/AssemblyUtil/assets/22202096/48760150-2257-4f9f-b035-a014d11595a0">

Reply:
Huh. What happens if you push that commit so the GHA tests run on it?

Reply:
Pass in 3 new files with parallelize flag = False without changing overflow logic. No complaint from codecov this time. 
huh...

Reply:
If parallelization is true, then new processes will be started and the generator will run in those new processes, right? Maybe codecov can't see those new processes?

Reply:
Looks like this is done :+1: